RecoTracker/LSTCore/src/alpaka/Kernels.h

```diff
@@ -274,6 +306,132 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::lst {
+            unsigned int jx = quadrupletModuleIndices_lowmod + jx1;
+
+            float eta2 = __H2F(quadruplets.eta()[jx]);
+            float phi2 = __H2F(quadruplets.phi()[jx]);
+            float dEta = alpaka::math::abs(acc, eta1 - eta2);
+            float dPhi = cms::alpakatools::deltaPhi(acc, phi1, phi2);
+            float displacedScore2 = quadruplets.displacedScore()[jx];
+            float fakeScore2 = quadruplets.fakeScore()[jx];
+            float score2 = displacedScore2 - fakeScore2;
+
+            if (dEta > 0.1f)
+              continue;
+
+            if (alpaka::math::abs(acc, dPhi) > 0.1f)
+              continue;
+
+            int nMatched = checkHitsT4(ix, jx, quadruplets);
+            const int minNHitsForDup_T4 = 5;
+            if (nMatched >= minNHitsForDup_T4) {
+              if (score1 >= score2) {
+                rmQuadrupletFromMemory(quadruplets, jx);
+              } else {
+                rmQuadrupletFromMemory(quadruplets, ix);
+              }
+            }
+          }
+        }
+      }
+    }
+  };
+
+  struct RemoveDupQuadrupletsBeforeTC {
+    ALPAKA_FN_ACC void operator()(Acc2D const& acc,
+                                  Quadruplets quadruplets,
+                                  QuadrupletsOccupancyConst quadrupletsOccupancy,
+                                  ObjectRangesConst ranges) const {
+      for (unsigned int lowmodIdx1 : cms::alpakatools::uniform_elements_y(acc, ranges.nEligibleT4Modules())) {
+        uint16_t lowmod1 = ranges.indicesOfEligibleT4Modules()[lowmodIdx1];
+        unsigned int nQuadruplets_lowmod1 = quadrupletsOccupancy.nQuadruplets()[lowmod1];
+        if (nQuadruplets_lowmod1 == 0)
+          continue;
+
+        unsigned int quadrupletModuleIndices_lowmod1 = ranges.quadrupletModuleIndices()[lowmod1];
+
+        for (unsigned int lowmodIdx2 :
+             cms::alpakatools::uniform_elements_x(acc, lowmodIdx1, ranges.nEligibleT4Modules())) {
+          uint16_t lowmod2 = ranges.indicesOfEligibleT4Modules()[lowmodIdx2];
+          unsigned int nQuadruplets_lowmod2 = quadrupletsOccupancy.nQuadruplets()[lowmod2];
+          if (nQuadruplets_lowmod2 == 0)
+            continue;
+
+          unsigned int quadrupletModuleIndices_lowmod2 = ranges.quadrupletModuleIndices()[lowmod2];
+
+          for (unsigned int ix1 = 0; ix1 < nQuadruplets_lowmod1; ix1 += 1) {
+            unsigned int ix = quadrupletModuleIndices_lowmod1 + ix1;
+            if ((quadruplets.isDup()[ix] & 1))
+              continue;
+
+            for (unsigned int jx1 = 0; jx1 < nQuadruplets_lowmod2; jx1++) {
+              unsigned int jx = quadrupletModuleIndices_lowmod2 + jx1;
+              if (ix == jx)
+                continue;
+
+              if ((quadruplets.isDup()[jx] & 1))
+                continue;
+
+              float eta1 = __H2F(quadruplets.eta()[ix]);
+              float phi1 = __H2F(quadruplets.phi()[ix]);
```

Comment:
```suggestion
              const float eta1 = __H2F(quadruplets.eta()[ix]);
              const float phi1 = __H2F(quadruplets.phi()[ix]);
```
also, move to the outer loop